scipy/spatial/tests/test_distance.py

```diff
@@ -631,31 +637,29 @@ def test_striding(self):
         assert_(X1_copy.flags.c_contiguous)
         assert_(X2_copy.flags.c_contiguous)
 
-        for metric in _METRICS_NAMES:
-            kwargs = dict()
-            if metric == 'minkowski':
-                kwargs['p'] = 1.23
-            Y1 = cdist(X1, X2, metric, **kwargs)
-            Y2 = cdist(X1_copy, X2_copy, metric, **kwargs)
-            # test that output is numerically equivalent
-            _assert_within_tol(Y1, Y2, eps, verbose > 2)
+        kwargs = dict()
+        if metric == 'minkowski':
+            kwargs['p'] = 1.23
+        Y1 = cdist(X1, X2, metric, **kwargs)
+        Y2 = cdist(X1_copy, X2_copy, metric, **kwargs)
+        # test that output is numerically equivalent
+        _assert_within_tol(Y1, Y2, eps, verbose > 2)
 
-    def test_cdist_refcount(self):
-        for metric in _METRICS_NAMES:
-            x1 = np.random.rand(10, 10)
-            x2 = np.random.rand(10, 10)
+    def test_cdist_refcount(self, metric):
+        x1 = np.random.rand(10, 10)
```

Comment:
We can also make these deterministic by using a seeded `RandomState`. Also `random` instead of `rand`

Reply:
`RandomState` is the [legacy](https://numpy.org/doc/stable/reference/random/legacy.html) API, see [overview](https://numpy.org/doc/stable/reference/random/) and [differences](https://numpy.org/doc/stable/reference/random/new-or-different.html#new-or-different)

The newer `default_rng -> Generator` is available since 1.17, which is I think the oldest numpy we're supporting anyway.

This is also veering quite far away from the intent of this PR, but if you want, I can do that.

Reply:
Yup, 1.16 was actually [dropped](https://github.com/scipy/scipy/pull/14655) by you. :)

Reply:
I know 😉

We had a few discussions on this when I made the switch. I also wanted to migrate everything to `Generator`, but it is still advices to use `RandomState` for the tests. The reason is that `RandomState` is not going to be removed and will stay there for this purpose while `Generator` is allowed to have its bite stream altered.

Reply:
Sure, but these tests are already running in an unseeded manner now. Surely the change from of default from `PCG64` to `PCG64DXSM` is irrelevant here? And if desired, I could already use the latter explicitly.

I just disagree with creating further tech debt if there is no strong reason to do it.

Reply:
Here we would reduce the technical debt by using `RandomState.random` not the other way around. It's not advised to use `Generator` there and as I said `RandomState` is kept in NumPy partially for that reason. Here for more information https://numpy.org/neps/nep-0019-rng-policy.html

Appart when we do hypothesis testing, we want to have reproducible tests to ensure we can exactly reproduce failing tests.

Reply:
> Here we would reduce the technical debt by using `RandomState.random` not the other way around.

Reducing it sure, but introducing new one (`RandomState` instead of `Generator`)

> It's not advised to use `Generator`

I disagree with that advice.

> we want to have reproducible tests to ensure we can exactly reproduce failing tests.

So reproducibility is good, fully agreed. But I actually very much dislike seeded randomness in tests, because we may just be continually testing a benign/pathological example. Hypothesis is indeed in the sweet spot here - generally random, but with clear seeds upon failure to reproduce it.

However, the tests in this module - as of right now - are not seeded, hence not dependent on such incredible reproducibility that we couldn't survive some very rare changes in the bitstream.

Reply:
> > It's not advised to use `Generator`
> 
> I disagree with that advice.

I got burned a few times here 😅 I also wanted at multiple time to go with `Generator` and was advised not to for tests. It's explained in the NEP as well.

That being said, I would personally validate both usage of `Generator` and `RandomState`. It should not really matter and in both cases would be an improvement 😃 

> So reproducibility is good, fully agreed. But I actually very much dislike seeded randomness in tests, because we may just be continually testing a benign/pathological example. Hypothesis is indeed in the sweet spot here - generally random, but with clear seeds upon failure to reproduce it.

It's always a difficult point. We are currently looking into adding some hypothesis testing for some stats, but it's tricky. There is a tradeoff to find because hypothesis testing is costly and what we do should bring value.

> However, the tests in this module - as of right now - are not seeded, hence not dependent on such incredible reproducibility that we couldn't survive some very rare changes in the bitstream.

There is this concern, but in this case it's something else. When there is a test failure which involves random numbers, if we don't have the background on the specific test we always wonder about the source of the issue and if the RNG is somehow involved. I've seen all cases happened and having a seed does help to identify faster the issue.

Reply:
> > @h-vetinari: I disagree with that advice.
>
> @tupui: I got burned a few times here 😅 I also wanted at multiple time to go with Generator and was advised not to for tests. It's explained in the NEP as well.

Don't get me wrong, I will go along with what is accepted practice here, but I don't have to agree with it. 😋

I read the NEP, and I think the quote here fits my point well (my emphasis):
> NEP19: The purpose of `RandomState` will be documented as providing certain fixed functionality for _**backwards compatibility**_ and stable numbers for the _**limited purpose**_ of unit testing, and not making whole programs reproducible across numpy versions.

The tests in question here _absolutely do not_ need to be stable with regard to randomness generated in previous numpy versions (indeed, the tests were unseeded up until now).

> @tupui: It's always a difficult point. We are currently looking into adding some hypothesis testing for some stats, but it's tricky. There is a tradeoff to find because hypothesis testing is costly and what we do should bring value.

Yeah, it will be a longterm effort. Still, I think it's good to reflect on why we want to (generally) use seeds - it helps with reproducibility, which helps with debugging, which helps with the scarcest resource: developer time (which matches with what you write in your last point). But _the tests themselves_ should not care about seed they get run with - otherwise there's a bug somewhere (which - however - we don't want to be confronted with intermittently)**.

Not annoying developers with flaky CI and irreproducible failures is a valid and worthwhile goal, locking the tests into a given seed & outcome is not (IMO).

** catching the latent bugs that are sleeping beneath fixed seeds (while still being respectful of developer time) is what hypothesis is very good at, but that'll be a long maturation process of the test suite (including knowing when/where to _not_ use it).

Reply:
I think the extent of the discussion here is a good indicator that we shouldn't include these changes in this PR. 🙃